### ISSUE TITLE
CardinalityEstimatorAdvancedTest.testCardinalityEstimatorFailure

### ISSUE THREAD

**Issue by hasancelik** [hazelcast/hazelcast#9017]
```
java.lang.AssertionError: expected:<2> but was:<0>
    at org.junit.Assert.fail(Assert.java:88)
    at org.junit.Assert.failNotEquals(Assert.java:834)
    at org.junit.Assert.assertEquals(Assert.java:645)
    at org.junit.Assert.assertEquals(Assert.java:631)
    at com.hazelcast.cardinality.CardinalityEstimatorAdvancedTest.testCardinalityEstimatorFailure(CardinalityEstimatorAdvancedTest.java:35)
```

https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/Hazelcast-3.x-OracleJDK1.7/com.hazelcast$hazelcast/623/testReport/junit/com.hazelcast.cardinality/CardinalityEstimatorAdvancedTest/testCardinalityEstimatorFailure/

**Comment by hasancelik** [comment]
https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/Hazelcast-3.x-OracleJDK8/952/

**Comment by hasancelik** [comment]
https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-OracleJDK1.6/com.hazelcast$hazelcast/1069/testReport/junit/com.hazelcast.cardinality/CardinalityEstimatorAdvancedTest/testCardinalityEstimatorFailure/

**Comment by hasancelik** [comment]
```
java.lang.AssertionError: expected:<2> but was:<0>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at com.hazelcast.cardinality.CardinalityEstimatorAdvancedTest.testCardinalityEstimatorFailure(CardinalityEstimatorAdvancedTest.java:35)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-ZuluJDK6/com.hazelcast$hazelcast/373/testReport/junit/com.hazelcast.cardinality/CardinalityEstimatorAdvancedTest/testCardinalityEstimatorFailure/

**Comment by hasancelik** [comment]
```
java.lang.AssertionError: expected:<2> but was:<0>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at com.hazelcast.cardinality.CardinalityEstimatorAdvancedTest.testCardinalityEstimatorFailure(CardinalityEstimatorAdvancedTest.java:35)
```

https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/Hazelcast-3.x-OracleJDK1.7/com.hazelcast$hazelcast/659/testReport/junit/com.hazelcast.cardinality/CardinalityEstimatorAdvancedTest/testCardinalityEstimatorFailure/

**Comment by hasancelik** [comment]
related to #9124 #9217

**Comment by hasancelik** [comment]
```
java.lang.AssertionError: expected:<2> but was:<0>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at com.hazelcast.cardinality.CardinalityEstimatorAdvancedTest.testCardinalityEstimatorFailure(CardinalityEstimatorAdvancedTest.java:35)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-IbmJDK1.8/com.hazelcast$hazelcast/404/testReport/junit/com.hazelcast.cardinality/CardinalityEstimatorAdvancedTest/testCardinalityEstimatorFailure/

**Comment by hasancelik** [comment]
```
java.lang.AssertionError: expected:<2> but was:<0>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at com.hazelcast.cardinality.CardinalityEstimatorAdvancedTest.testCardinalityEstimatorFailure(CardinalityEstimatorAdvancedTest.java:35)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-OracleJDK1.7/com.hazelcast$hazelcast/682/testReport/junit/com.hazelcast.cardinality/CardinalityEstimatorAdvancedTest/testCardinalityEstimatorFailure/

**Comment by hasancelik** [comment]
```
java.lang.AssertionError: expected:<2> but was:<0>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at com.hazelcast.cardinality.CardinalityEstimatorAdvancedTest.testCardinalityEstimatorFailure(CardinalityEstimatorAdvancedTest.java:35)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-IbmJDK1.7/com.hazelcast$hazelcast/1120/testReport/junit/com.hazelcast.cardinality/CardinalityEstimatorAdvancedTest/testCardinalityEstimatorFailure/

**Comment by hasancelik** [comment]
```
java.lang.AssertionError: expected:<2> but was:<0>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at com.hazelcast.cardinality.CardinalityEstimatorAdvancedTest.testCardinalityEstimatorFailure(CardinalityEstimatorAdvancedTest.java:35)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-OracleJDK1.6/com.hazelcast$hazelcast/1134/testReport/junit/com.hazelcast.cardinality/CardinalityEstimatorAdvancedTest/testCardinalityEstimatorFailure/

**Comment by hasancelik** [comment]
```
org.junit.runners.model.TestTimedOutException: test timed out after 300000 milliseconds
	at org.apache.log4j.Category.callAppenders(Category.java:201)
	at org.apache.log4j.Category.forcedLog(Category.java:388)
	at org.apache.log4j.Category.log(Category.java:823)
	at com.hazelcast.logging.Log4jFactory$Log4jLogger.log(Log4jFactory.java:50)
	at com.hazelcast.logging.LoggingServiceImpl$DefaultLogger.log(LoggingServiceImpl.java:167)
	at com.hazelcast.logging.LoggingServiceImpl$DefaultLogger.log(LoggingServiceImpl.java:155)
	at com.hazelcast.logging.AbstractLogger.finest(AbstractLogger.java:30)
	at com.hazelcast.spi.impl.servicemanager.impl.ServiceManagerImpl.shutdownService(ServiceManagerImpl.java:300)
	at com.hazelcast.spi.impl.servicemanager.impl.ServiceManagerImpl.shutdown(ServiceManagerImpl.java:293)
	at com.hazelcast.spi.impl.NodeEngineImpl.shutdown(NodeEngineImpl.java:489)
	at com.hazelcast.instance.Node.shutdownServices(Node.java:488)
	at com.hazelcast.instance.Node.shutdown(Node.java:462)
	at com.hazelcast.instance.LifecycleServiceImpl.shutdown(LifecycleServiceImpl.java:98)
	at com.hazelcast.instance.LifecycleServiceImpl.shutdown(LifecycleServiceImpl.java:81)
	at com.hazelcast.instance.HazelcastInstanceProxy.shutdown(HazelcastInstanceProxy.java:290)
	at com.hazelcast.cardinality.CardinalityEstimatorAdvancedTest.testCardinalityEstimatorFailure(CardinalityEstimatorAdvancedTest.java:37)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x/com.hazelcast$hazelcast/6279/testReport/junit/com.hazelcast.cardinality/CardinalityEstimatorAdvancedTest/testCardinalityEstimatorFailure/

**Comment by hasancelik** [comment]
```
java.lang.AssertionError: expected:<2> but was:<0>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at com.hazelcast.cardinality.CardinalityEstimatorAdvancedTest.testCardinalityEstimatorFailure(CardinalityEstimatorAdvancedTest.java:35)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-IbmJDK1.8/com.hazelcast$hazelcast/461/testReport/junit/com.hazelcast.cardinality/CardinalityEstimatorAdvancedTest/testCardinalityEstimatorFailure/

**Comment by tkountis** [comment]
Functionality seems to be correct, I don't see any race condition that could cause this. System seems to be going in overdrive, big delays in operations, big cluster time shifts and many migrations from which many aren't successful due to Nodes being still in STARTING mode. At some point, partitions are lost completely, which leads to the wrong figures. This is the same behaviour with both #9124  and #9217. 

I checked system load on the diagnostic logs of the last build, and load was ok, no reason to have caused these delays. I don't know what caused this behaviour, but it also seems related to #9202. I would close the issue.

**Comment by hasancelik** [comment]
```
java.lang.NullPointerException: null
	at com.hazelcast.test.HazelcastTestSupport.getClusterService(HazelcastTestSupport.java:182)
	at com.hazelcast.test.HazelcastTestSupport.getAddress(HazelcastTestSupport.java:206)
	at com.hazelcast.test.HazelcastTestSupport.assertAllInSafeState(HazelcastTestSupport.java:667)
	at com.hazelcast.test.HazelcastTestSupport$4.run(HazelcastTestSupport.java:645)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1016)
	at com.hazelcast.test.HazelcastTestSupport.waitAllForSafeState(HazelcastTestSupport.java:643)
	at com.hazelcast.test.HazelcastTestSupport.waitAllForSafeState(HazelcastTestSupport.java:639)
	at com.hazelcast.test.HazelcastTestSupport.waitAllForSafeState(HazelcastTestSupport.java:635)
	at com.hazelcast.cardinality.CardinalityEstimatorAdvancedTest.testCardinalityEstimatorFailure(CardinalityEstimatorAdvancedTest.java:34)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-IbmJDK1.7/com.hazelcast$hazelcast/1156/testReport/junit/com.hazelcast.cardinality/CardinalityEstimatorAdvancedTest/testCardinalityEstimatorFailure/

**Comment by jerrinot** [comment]
closed by #9661
